relayer/checkpoint.go

```diff
@@ -0,0 +1,210 @@
+// Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package relayer
+
+import (
+	"container/heap"
+	"strconv"
+	"sync"
+
+	"github.com/ava-labs/avalanchego/utils/logging"
+	"github.com/ava-labs/awm-relayer/database"
+	"github.com/ava-labs/awm-relayer/utils"
+	"go.uber.org/zap"
+)
+
+//
+// checkpointManager commits keys to be written to the database in a thread safe manner.
+//
+
+type checkpointManager struct {
+	logger                   logging.Logger
+	database                 database.RelayerDatabase
+	writeSignal              chan struct{}
+	relayerID                database.RelayerID
+	queuedHeightsAndMessages map[uint64]*messageCounter
+	committedHeight          uint64
+	lock                     *sync.RWMutex
+	pendingCommits           *utils.UInt64Heap
+	finished                 chan uint64
+}
+
+func newCheckpointManager(logger logging.Logger, database database.RelayerDatabase, writeSignal chan struct{}, relayerID database.RelayerID) *checkpointManager {
+	h := &utils.UInt64Heap{}
+	heap.Init(h)
+	return &checkpointManager{
+		logger:                   logger,
+		database:                 database,
+		writeSignal:              writeSignal,
+		relayerID:                relayerID,
+		queuedHeightsAndMessages: make(map[uint64]*messageCounter),
+		lock:                     &sync.RWMutex{},
+		pendingCommits:           h,
+		finished:                 make(chan uint64),
+	}
+}
+
+func (cm *checkpointManager) run() {
+	go cm.listenForFinishedRelays()
+	go cm.listenForWriteSignal()
+}
+
+func (cm *checkpointManager) writeToDatabase() {
+	cm.lock.RLock()
+	defer cm.lock.RUnlock()
+	// Ensure we're not writing the default value
+	if cm.committedHeight == 0 {
+		return
+	}
+	storedHeight, err := getLatestProcessedBlockHeight(cm.database, cm.relayerID)
+	if err != nil && !database.IsKeyNotFoundError(err) {
+		cm.logger.Error(
+			"Failed to get latest processed block height",
+			zap.Error(err),
+			zap.String("relayerID", cm.relayerID.ID.String()),
+		)
+		return
+	}
+	if storedHeight >= cm.committedHeight {
+		return
+	}
+	cm.logger.Debug(
+		"Writing height",
+		zap.Uint64("height", cm.committedHeight),
+		zap.String("relayerID", cm.relayerID.ID.String()),
+	)
+	err = cm.database.Put(cm.relayerID.ID, database.LatestProcessedBlockKey, []byte(strconv.FormatUint(cm.committedHeight, 10)))
+	if err != nil {
+		cm.logger.Error(
+			"Failed to write latest processed block height",
+			zap.Error(err),
+			zap.String("relayerID", cm.relayerID.ID.String()),
+		)
+		return
+	}
+}
+
+func (cm *checkpointManager) listenForWriteSignal() {
+	for range cm.writeSignal {
+		cm.writeToDatabase()
+	}
+}
+
+func (cm *checkpointManager) incrementFinishedCounter(height uint64) {
+	cm.lock.Lock()
+	defer cm.lock.Unlock()
+	counter, ok := cm.queuedHeightsAndMessages[height]
+	if !ok {
```

Comment:
should this case ever happen? Confirming when this happens right now we just return and wait for the next finished block height?

Reply:
This is actually expected to happen when processing manual Warp messages. In that case, the message is not associated with a block height, so it does not go through the usual prepare-complete step as normal Warp messages. I've demoted this to a debug log and added a comment explaining this.

And to clarify, when this case is hit, we no-op and continue listening for finished relay signals.